File: server/internal/harmonica/engine.go
package harmonica

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"unicode/utf8"
)

// Supported session languages
const (
	LangZhHant = "zh-Hant"
	LangEn     = "en"
)

// Default constraints aligned with Pocket Harmonica
const (
	DefaultMaxTurns        = 6
	MaxTurnsCap            = 10
	DefaultMaxParticipants = 20
	MaxParticipantsCap     = 100
	PromptMaxBytes         = 6000
)

// SessionSettings holds configuration for an interview session.
type SessionSettings struct {
	Topic           string   `json:"topic"`
	Goal            string   `json:"goal"`
	Context         string   `json:"context,omitempty"`
	Critical        string   `json:"critical,omitempty"`
	Questions       []string `json:"questions"`
	Language        string   `json:"language"`
	MaxTurns        int      `json:"maxTurns"`
	MaxParticipants int      `json:"maxParticipants"`
	AskAlias        bool     `json:"askAlias"`
}

// Message is an individual dialogue entry.
type Message struct {
	Seq  int    `json:"seq"`
	Role string `json:"role"` // "interviewer" | "participant"
	Text string `json:"text"`
	At   int64  `json:"at"`
}

// Conversation records a participant's complete dialogue track.
type Conversation struct {
	Participant int       `json:"participant"`
	Alias       string    `json:"alias"`
	Turns       int       `json:"turns"`
	Done        bool      `json:"done"`
	Messages    []Message `json:"messages"`
	UpdatedAt   int64     `json:"updatedAt"`
}

// PublicSession is the metadata view of an interview.
type PublicSession struct {
	SessionID       string   `json:"sessionId"`
	Topic           string   `json:"topic"`
	Goal            string   `json:"goal"`
	Context         string   `json:"context,omitempty"`
	Critical        string   `json:"critical,omitempty"`
	Questions       []string `json:"questions"`
	Language        string   `json:"language"`
	Status          string   `json:"status"` // "open" | "closed" | "deleted"
	MaxTurns        int      `json:"maxTurns"`
	MaxParticipants int      `json:"maxParticipants"`
	AskAlias        bool     `json:"askAlias"`
	Participants    int      `json:"participants"`
	Completed       int      `json:"completed"`
	CreatedAt       int64    `json:"createdAt"`
	UpdatedAt       int64    `json:"updatedAt"`
	Model           string   `json:"model"`
	AdminToken      string   `json:"adminToken,omitempty"` // only present in creation response
}

// HostConversationSummary is a summary view for the host dashboard.
type HostConversationSummary struct {
	Participant int    `json:"participant"`
	Alias       string `json:"alias"`
	Turns       int    `json:"turns"`
	Done        bool   `json:"done"`
	Messages    int    `json:"messages"`
	UpdatedAt   int64  `json:"updatedAt"`
	LastMessage string `json:"lastMessage,omitempty"`
}

// HostView is the full dashboard view for organizers.
type HostView struct {
	PublicSession
	Conversations []Conversation            `json:"conversations"`
	Summaries     []HostConversationSummary `json:"summaries"`
}

// NormalizeSettings validates and normalizes raw interview settings.
func NormalizeSettings(s SessionSettings) (SessionSettings, error) {
	topic := CleanLine(s.Topic, 120)
	if topic == "" {
		return s, fmt.Errorf("先幫這輪訪談取一個名字 (topic is required)")
	}

	goal := CleanText(s.Goal, 500)
	if goal == "" {
		return s, fmt.Errorf("請說明希望這輪訪談理解什麼 (goal is required)")
	}

	questions := make([]string, 0, len(s.Questions))
	for _, q := range s.Questions {
		cleaned := CleanLine(q, 240)
		if cleaned != "" {
			questions = append(questions, cleaned)
		}
		if len(questions) >= 8 {
			break
		}
	}
	if len(questions) == 0 {
		return s, fmt.Errorf("至少給一個起始問題 (at least one question is required)")
	}

	lang := LangZhHant
	if strings.ToLower(s.Language) == "en" {
		lang = LangEn
	}

	maxTurns := s.MaxTurns
	if maxTurns < 2 {
		maxTurns = DefaultMaxTurns
	}
	if maxTurns > MaxTurnsCap {
		maxTurns = MaxTurnsCap
	}

	maxParticipants := s.MaxParticipants
	if maxParticipants < 1 {
		maxParticipants = DefaultMaxParticipants
	}
	if maxParticipants > MaxParticipantsCap {
		maxParticipants = MaxParticipantsCap
	}

	return SessionSettings{
		Topic:           topic,
		Goal:            goal,
		Context:         CleanText(s.Context, 1000),
		Critical:        CleanText(s.Critical, 500),
		Questions:       questions,
		Language:        lang,
		MaxTurns:        maxTurns,
		MaxParticipants: maxParticipants,
		AskAlias:        s.AskAlias,
	}, nil
}

// OpeningMessage produces the zero-token greeting and first question.
func OpeningMessage(s SessionSettings) string {
	first := ""
	if len(s.Questions) > 0 {
		first = s.Questions[0]
	}

	if s.Language == LangEn {
		return fmt.Sprintf("Thanks for joining this conversation about \"%s\". %s I'll ask a few questions and follow up on what you say; there are no right answers, and you can stop at any time. To start: %s", s.Topic, s.Goal, first)
	}

	return fmt.Sprintf("謝謝你參加這輪關於「%s」的對話。%s我會問幾個問題，並依你說的內容追問；沒有標準答案，隨時可以停。先從這個開始：%s", s.Topic, s.Goal, first)
}

// BuildSystemPrompt crafts Pocket Harmonica's warm, neutral interviewer persona.
func BuildSystemPrompt(s SessionSettings) string {
	var lines []string

	lines = append(lines, "You are a warm, neutral interviewer for a public consultation. You listen, ask one question at a time, and follow up on concrete experiences, reasons and trade-offs the participant mentions.")
	lines = append(lines, fmt.Sprintf("Topic: %s", s.Topic))
	lines = append(lines, fmt.Sprintf("What the organizer wants to understand: %s", s.Goal))

	if s.Context != "" {
		lines = append(lines, fmt.Sprintf("Background: %s", s.Context))
	}
	if s.Critical != "" {
		lines = append(lines, fmt.Sprintf("Voices or issues that must not be missed: %s", s.Critical))
	}

	var qList []string
	for i, q := range s.Questions {
		qList = append(qList, fmt.Sprintf("%d. %s", i+1, q))
	}
	lines = append(lines, fmt.Sprintf("Starter questions, in order: %s", strings.Join(qList, " ")))

	lines = append(lines, "Rules: never argue, evaluate or persuade; do not add facts of your own; reflect briefly what you heard, then ask exactly one question (a follow-up on something specific, or the next starter question when the current one feels answered). Keep each reply under 90 words.")

	if s.Language == LangEn {
		lines = append(lines, "Write in English. Reply with plain text only: no lists, no markdown, no preamble.")
	} else {
		lines = append(lines, "Write in Traditional Chinese as used in Taiwan (zh-Hant-TW). Reply with plain text only: no lists, no markdown, no preamble.")
	}

	return strings.Join(lines, "\n")
}

// BuildUserPrompt formats conversation history and instructions for the current turn.
func BuildUserPrompt(s SessionSettings, history []Message, latest string, turn int, asked int) string {
	isFinal := turn >= s.MaxTurns
	remaining := []string{}
	if asked < len(s.Questions) {
		remaining = s.Questions[asked:]
	}

	var tailLines []string
	tailLines = append(tailLines, fmt.Sprintf("Turn %d of %d.", turn, s.MaxTurns))

	if isFinal {
		tailLines = append(tailLines, "FINAL TURN: thank the participant, reflect the most important thing they said in one sentence, and close without asking anything.")
	} else if len(remaining) > 0 {
		tailLines = append(tailLines, fmt.Sprintf("Next starter question to ask when the current one feels answered: %s", remaining[0]))
	} else {
		tailLines = append(tailLines, "All starter questions have been asked; follow up on what matters most to them, or gently close if they seem done.")
	}

	tailLines = append(tailLines, fmt.Sprintf("Participant just said: %s", latest))
	tail := strings.Join(tailLines, "\n")

	budget := PromptMaxBytes - len([]byte(tail)) - 40
	if budget < 0 {
		budget = 500
	}

	var transcript []string
	used := 0
	for i := len(history) - 1; i >= 0; i-- {
		m := history[i]
		role := "Participant"
		if m.Role == "interviewer" {
			role = "Interviewer"
		}
		line := fmt.Sprintf("%s: %s", role, m.Text)
		bLen := len([]byte(line)) + 1
		if used+bLen > budget {
			break
		}
		transcript = append([]string{line}, transcript...)
		used += bLen
	}

	historyStr := "(none)"
	if len(transcript) > 0 {
		historyStr = strings.Join(transcript, "\n")
	}

	return fmt.Sprintf("Conversation so far:\n%s\n\n%s", historyStr, tail)
}

// CleanReply scrubs unwanted reasoning tokens, markdown symbols, and interviewer labels.
func CleanReply(text string, maxLen int) string {
	// Strip <think>...</think> or reasoning tags if present
	reTag := regexp.MustCompile(`(?s)<(?:think|reasoning)>.*?</(?:think|reasoning)>`)
	cleaned := reTag.ReplaceAllString(text, "")

	// Strip markdown blocks
	reCode := regexp.MustCompile("(?s)```.*?```")
	cleaned = reCode.ReplaceAllString(cleaned, "")

	// Strip leading role labels e.g. "Interviewer:" or "訪談者："
	reLabel := regexp.MustCompile(`(?i)^\s*(Interviewer|訪談者|AI|助手)[:：]\s*`)
	cleaned = reLabel.ReplaceAllString(cleaned, "")

	// Normalize spaces
	reSpace := regexp.MustCompile(`\s+`)
	cleaned = strings.TrimSpace(reSpace.ReplaceAllString(cleaned, " "))

	if maxLen <= 0 {
		maxLen = 600
	}

	if utf8.RuneCountInString(cleaned) > maxLen {
		runes := []rune(cleaned)
		cleaned = string(runes[:maxLen])
	}

	if cleaned == "" {
		cleaned = "謝謝你的分享，能再多聊聊具體的考量嗎？"
	}

	return cleaned
}

// CleanLine removes control characters and trims string.
func CleanLine(v string, maxLen int) string {
	cleaned := strings.Map(func(r rune) rune {
		if r < 32 && r != '\t' {
			return ' '
		}
		return r
	}, v)
	cleaned = strings.TrimSpace(strings.Join(strings.Fields(cleaned), " "))
	if maxLen > 0 && utf8.RuneCountInString(cleaned) > maxLen {
		runes := []rune(cleaned)
		cleaned = string(runes[:maxLen])
	}
	return cleaned
}

// CleanText normalizes multi-line text.
func CleanText(v string, maxLen int) string {
	cleaned := strings.ReplaceAll(v, "\r\n", "\n")
	cleaned = strings.ReplaceAll(cleaned, "\r", "\n")
	cleaned = strings.TrimSpace(cleaned)
	if maxLen > 0 && utf8.RuneCountInString(cleaned) > maxLen {
		runes := []rune(cleaned)
		cleaned = string(runes[:maxLen])
	}
	return cleaned
}

// FormatTTTCCsv exports all conversations in Talk to the City CSV format: id,interview,comment
func FormatTTTCCsv(session PublicSession, conversations []Conversation) ([]byte, error) {
	var buf bytes.Buffer
	writer := csv.NewWriter(&buf)

	// TTTC standard headers
	if err := writer.Write([]string{"id", "interview", "comment"}); err != nil {
		return nil, err
	}

	msgID := 1
	for _, conv := range conversations {
		alias := conv.Alias
		if alias == "" {
			alias = fmt.Sprintf("Participant_%d", conv.Participant)
		}
		interviewName := fmt.Sprintf("%s (%s)", session.Topic, alias)

		for _, msg := range conv.Messages {
			if msg.Role == "participant" && strings.TrimSpace(msg.Text) != "" {
				row := []string{
					fmt.Sprintf("%d", msgID),
					interviewName,
					msg.Text,
				}
				if err := writer.Write(row); err != nil {
					return nil, err
				}
				msgID++
			}
		}
	}

	writer.Flush()
	return buf.Bytes(), writer.Error()
}

// FormatTranscriptsJSON exports all conversations in full structured JSON format.
func FormatTranscriptsJSON(session PublicSession, conversations []Conversation) ([]byte, error) {
	out := map[string]interface{}{
		"session":       session,
		"exportedAt":    session.UpdatedAt,
		"conversations": conversations,
	}
	return json.MarshalIndent(out, "", "  ")
}

File: web/src/interview/engine.ts
import {
  HarmonicaSession,
  HarmonicaMessage,
  HarmonicaConversation,
  HarmonicaHostView,
  HarmonicaSessionRef,
  CreateHarmonicaSessionPayload,
} from '../types';

export const PROMPT_MAX_BYTES = 6000;
export const DEFAULT_MAX_TURNS = 6;
export const MAX_TURNS_CAP = 10;
export const DEFAULT_MAX_PARTICIPANTS = 20;

const encoder = new TextEncoder();
const decoder = new TextDecoder();

export function utf8ByteLength(text: string): number {
  return encoder.encode(text).length;
}

export function truncateUtf8(text: string, maxBytes: number): string {
  if (maxBytes <= 0) return '';
  const bytes = encoder.encode(text);
  if (bytes.length <= maxBytes) return text;
  let end = maxBytes;
  while (end > 0 && ((bytes[end] ?? 0) & 0b1100_0000) === 0b1000_0000) end -= 1;
  return decoder.decode(bytes.slice(0, end));
}

export function cleanLine(value: unknown, max = 240): string {
  if (typeof value !== 'string') return '';
  return value
    .replace(/[\u0000-\u0008\u000b\u000c\u000e-\u001f]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
    .slice(0, max);
}

export function cleanText(value: unknown, max = 1000): string {
  if (typeof value !== 'string') return '';
  return value
    .replace(/[\u0000-\u0008\u000b\u000c\u000e-\u001f]/g, '')
    .replace(/\r\n?/g, '\n')
    .replace(/[\t ]+/g, ' ')
    .replace(/\n{3,}/g, '\n\n')
    .trim()
    .slice(0, max);
}

export function cleanReply(text: string, max = 600): string {
  const cleaned = text
    .replace(/<think>[\s\S]*?<\/think>/gi, '')
    .replace(/```[\s\S]*?```/g, '')
    .replace(/^\s*(Interviewer|訪談者|AI|助手)[:：]\s*/i, '')
    .replace(/\s+/g, ' ')
    .trim();
  return truncateUtf8(cleaned, max * 3).slice(0, max) || '謝謝你分享，可以再多聊聊具體的考量嗎？';
}

export function normalizeSettings(raw: Partial<CreateHarmonicaSessionPayload>): CreateHarmonicaSessionPayload {
  const topic = cleanLine(raw.topic, 120);
  const goal = cleanText(raw.goal, 500);
  if (!topic) throw new Error('先幫這輪訪談取一個名字');
  if (!goal) throw new Error('請說明希望這輪訪談理解什麼');

  const questions = (Array.isArray(raw.questions) ? raw.questions : [])
    .map((q) => cleanLine(q, 240))
    .filter(Boolean)
    .slice(0, 8);
  if (questions.length === 0) throw new Error('至少給一個起始問題');

  const language = raw.language === 'en' ? 'en' : 'zh-Hant';
  const maxTurns = Math.min(MAX_TURNS_CAP, Math.max(2, raw.maxTurns || DEFAULT_MAX_TURNS));
  const maxParticipants = Math.min(100, Math.max(1, raw.maxParticipants || DEFAULT_MAX_PARTICIPANTS));

  return {
    topic,
    goal,
    context: cleanText(raw.context, 1000),
    critical: cleanText(raw.critical, 500),
    questions,
    language,
    maxTurns,
    maxParticipants,
    askAlias: raw.askAlias !== false,
  };
}

export function openingMessage(settings: CreateHarmonicaSessionPayload): string {
  const first = settings.questions[0] ?? '';
  if (settings.language === 'en') {
    return `Thanks for joining this conversation about "${settings.topic}". ${settings.goal} I'll ask a few questions and follow up on what you say; there are no right answers, and you can stop at any time. To start: ${first}`;
  }
  return `謝謝你參加這輪關於「${settings.topic}」的對話。${settings.goal}我會問幾個問題，並依你說的內容追問；沒有標準答案，隨時可以停。先從這個開始：${first}`;
}

export function buildSystemPrompt(settings: CreateHarmonicaSessionPayload): string {
  const lines = [
    'You are a warm, neutral interviewer for a public consultation. You listen, ask one question at a time, and follow up on concrete experiences, reasons and trade-offs the participant mentions.',
    `Topic: ${settings.topic}`,
    `What the organizer wants to understand: ${settings.goal}`,
    settings.context ? `Background: ${settings.context}` : '',
    settings.critical ? `Voices or issues that must not be missed: ${settings.critical}` : '',
    `Starter questions, in order: ${settings.questions.map((q, i) => `${i + 1}. ${q}`).join(' ')}`,
    'Rules: never argue, evaluate or persuade; do not add facts of your own; reflect briefly what you heard, then ask exactly one question (a follow-up on something specific, or the next starter question when the current one feels answered). Keep each reply under 90 words.',
    settings.language === 'en'
      ? 'Write in English. Reply with plain text only: no lists, no markdown, no preamble.'
      : 'Write in Traditional Chinese as used in Taiwan (zh-Hant-TW). Reply with plain text only: no lists, no markdown, no preamble.',
  ];
  return lines.filter(Boolean).join('\n');
}

export function buildUserPrompt(
  settings: CreateHarmonicaSessionPayload,
  messages: HarmonicaMessage[],
  latest: string,
  turn: number,
  asked: number
): string {
  const remaining = settings.questions.slice(asked);
  const isFinal = turn >= settings.maxTurns;
  const tail = [
    `Turn ${turn} of ${settings.maxTurns}.`,
    isFinal
      ? 'FINAL TURN: thank the participant, reflect the most important thing they said in one sentence, and close without asking anything.'
      : remaining.length > 0
      ? `Next starter question to ask when the current one feels answered: ${remaining[0]}`
      : 'All starter questions have been asked; follow up on what matters most to them, or gently close if they seem done.',
    `Participant just said: ${latest}`,
  ].join('\n');

  const budget = PROMPT_MAX_BYTES - utf8ByteLength(tail) - 40;
  const transcript: string[] = [];
  let used = 0;

  for (let i = messages.length - 1; i >= 0; i--) {
    const m = messages[i];
    const line = `${m.role === 'interviewer' ? 'Interviewer' : 'Participant'}: ${m.text}`;
    const bytes = utf8ByteLength(line) + 1;
    if (used + bytes > budget) break;
    transcript.unshift(line);
    used += bytes;
  }

  return `Conversation so far:\n${transcript.length > 0 ? transcript.join('\n') : '(none)'}\n\n${tail}`;
}

export function generateLocalReply(
  settings: CreateHarmonicaSessionPayload,
  _latestText: string,
  turn: number,
  isFinal: boolean
): string {
  if (isFinal) {
    if (settings.language === 'en') {
      return 'Thank you very much for taking the time to share your perspective. Your input has been recorded and will be summarized for public deliberation.';
    }
    return '非常感謝您撥冗分享寶貴的想法。您的多元觀點已被客觀記錄，將作為公共審議與政策評估的重要參考。訪談在此圓滿結束！';
  }

  // If next question exists, ask it
  if (turn - 1 < settings.questions.length) {
    const nextQ = settings.questions[turn - 1];
    if (settings.language === 'en') {
      return `Thank you for sharing that. Following up on your points: ${nextQ}`;
    }
    return `謝謝您的具體說明。延續剛才的話題，想請教您：${nextQ}`;
  }

  if (settings.language === 'en') {
    return 'That makes a lot of sense. Could you elaborate a bit more on what specific conditions or priorities matter most to you?';
  }
  return '理解您的考量。在剛才提到的內容中，您認為最重要的優先順序或關鍵配套會是什麼？';
}

export function formatTTTCCsv(session: HarmonicaSession, conversations: HarmonicaConversation[]): string {
  const lines = ['id,interview,comment'];
  let id = 1;

  for (const conv of conversations) {
    const alias = conv.alias || `Participant_${conv.participant}`;
    const interview = `"${session.topic} (${alias})".replace(/"/g, '""')`;
    for (const msg of conv.messages) {
      if (msg.role === 'participant' && msg.text.trim()) {
        const escapedComment = `"${msg.text.replace(/"/g, '""')}"`;
        lines.push(`${id},${interview},${escapedComment}`);
        id++;
      }
    }
  }

  return lines.join('\n');
}

export function formatTranscriptsJson(session: HarmonicaSession, conversations: HarmonicaConversation[]): string {
  return JSON.stringify(
    {
      session,
      exportedAt: Date.now(),
      conversations,
    },
    null,
    2
  );
}

// Client-side LocalStorage repository for native Harmonica sessions
export class HarmonicaLocalStore {
  private static PREFIX = 'topos:native_harmonica';

  static getTopicSessions(topicId: string): HarmonicaSessionRef[] {
    try {
      const raw = localStorage.getItem(`${this.PREFIX}:topic_sessions:${topicId}`);
      if (raw) {
        const parsed = JSON.parse(raw);
        if (Array.isArray(parsed) && parsed.length > 0) return parsed;
      }
    } catch (_) {}

    // Pre-seeded defaults
    if (topicId === 'nuclear4') {
      return [
        {
          sessionId: '6o0ryapw2o',
          title: '核電重啟公眾訪談',
          goal: '測試與收集核電重啟之條件、顧慮與多元考量',
          url: '/s/6o0ryapw2o',
          isDefault: true,
          createdAt: 1790167501,
        },
      ];
    }

    if (topicId === 'sports-station') {
      return [
        {
          sessionId: 'sports-int-1',
          title: '運動驛站公眾訪談',
          goal: '探討公眾對捷運站盥洗寄物設施之需求、衛生管理與預算自償考量',
          url: '/s/sports-int-1',
          isDefault: true,
          createdAt: 1790167501,
        },
      ];
    }

    if (topicId === 'control-yuan') {
      return [
        {
          sessionId: 'cy-voices-01',
          title: '監察與考試權公眾訪談',
          goal: '理解民眾對監察權、考試權存廢的經驗、擔憂與期待。',
          url: '/s/cy-voices-01',
          isDefault: true,
          createdAt: 1790170000,
        },
      ];
    }

    return [];
  }

  static addTopicSession(topicId: string, ref: HarmonicaSessionRef): void {
    const list = this.getTopicSessions(topicId);
    const filtered = list.filter((s) => s.sessionId !== ref.sessionId);
    const updated = [ref, ...filtered];
    try {
      localStorage.setItem(`${this.PREFIX}:topic_sessions:${topicId}`, JSON.stringify(updated));
    } catch (_) {}
  }

  static getSession(sessionId: string): HarmonicaSession | null {
    try {
      const raw = localStorage.getItem(`${this.PREFIX}:session:${sessionId}`);
      if (raw) return JSON.parse(raw);
    } catch (_) {}

    // Pre-seeded defaults
    if (sessionId === '6o0ryapw2o') {
      return {
        sessionId: '6o0ryapw2o',
        topic: '核電重啟公眾訪談',
        goal: '測試與收集核電重啟之條件、顧慮與多元考量。',
        context: '探討台灣能源轉型與核四重啟之關鍵爭點與各方考量。',
        questions: [
          '對於核電重啟，你的基本立場是什麼？最在意的是哪一點？',
          '如果真的要重啟，你認為必須先滿足哪些條件？',
          '核廢料該怎麼處理，你心中有能接受的做法嗎？',
        ],
        language: 'zh-Hant',
        status: 'open',
        maxTurns: 6,
        maxParticipants: 50,
        askAlias: true,
        participants: 2,
        completed: 1,
        model: '@cf/google/gemma-4-26b-a4b-it',
        adminToken: 'admin-nuclear4-topos',
      };
    }

    if (sessionId === 'sports-int-1') {
      return {
        sessionId: 'sports-int-1',
        topic: '運動驛站市民訪談',
        goal: '探討公眾對捷運站盥洗寄物設施之需求、衛生管理與預算自償考量',
        context: '台北市長選舉市政政見討論，平衡運動友善與公共治安。',
        questions: [
          '您平日在捷運站或登山口有淋浴更衣的需求嗎？',
          '您認為此類設施應由公帑全額負擔，或採使用者付費？',
          '如何兼顧公共衛生、隱私防偷拍與深夜維安？',
        ],
        language: 'zh-Hant',
        status: 'open',
        maxTurns: 6,
        maxParticipants: 50,
        askAlias: true,
        participants: 0,
        completed: 0,
        model: '@cf/google/gemma-4-26b-a4b-it',
        adminToken: 'admin-sports-topos',
      };
    }

    if (sessionId === 'cy-voices-01') {
      return {
        sessionId: 'cy-voices-01',
        topic: '監察與考試權公眾訪談',
        goal: '理解民眾對監察權、考試權存廢的經驗、擔憂與期待。',
        context: '五權憲法架構在現代民主體制的運作困境、彈劾與調查權歸屬。',
        questions: [
          '你對監察院或考試院的印象是什麼？有沒有具體經驗？',
          '如果彈劾與調查權移到國會，你最擔心什麼？',
          '什麼樣的制度設計，會讓你覺得監督是可信的？',
        ],
        language: 'zh-Hant',
        status: 'open',
        maxTurns: 5,
        maxParticipants: 50,
        askAlias: true,
        participants: 0,
        completed: 0,
        model: '@cf/google/gemma-4-26b-a4b-it',
        adminToken: 'admin-cy-topos',
      };
    }

    return null;
  }

  static saveSession(session: HarmonicaSession): void {
    try {
      localStorage.setItem(`${this.PREFIX}:session:${session.sessionId}`, JSON.stringify(session));
    } catch (_) {}
  }

  static getConversations(sessionId: string): HarmonicaConversation[] {
    try {
      const raw = localStorage.getItem(`${this.PREFIX}:conversations:${sessionId}`);
      if (raw) return JSON.parse(raw);
    } catch (_) {}

    if (sessionId === '6o0ryapw2o') {
      return [
        {
          participant: 1,
          alias: '新北通勤族',
          turns: 2,
          done: false,
          messages: [
            { role: 'interviewer', text: '你好，新北通勤族！謝謝你願意聊聊「核電重啟公眾訪談」。\n\n第一個問題：對於核電重啟，你的基本立場是什麼？最在意的是哪一點？' },
            { role: 'participant', text: '我其實不反對，但住在北部，最擔心的是真的出事的時候疏散怎麼辦。' },
            { role: 'interviewer', text: '你提到「疏散怎麼辦」——可以多說一點嗎？是什麼經驗或資訊讓你這樣想？' },
            { role: 'participant', text: '福島那時候的畫面吧。新北人口這麼多，塞車就走不了。' },
            { role: 'interviewer', text: '了解，謝謝你。換個角度：如果真的要重啟，你認為必須先滿足哪些條件？' },
          ],
        },
        {
          participant: 2,
          alias: '南部家長',
          turns: 6,
          done: true,
          messages: [
            { role: 'interviewer', text: '你好，南部家長！謝謝你願意聊聊「核電重啟公眾訪談」。\n\n第一個問題：對於核電重啟，你的基本立場是什麼？最在意的是哪一點？' },
            { role: 'participant', text: '我比較支持，因為中南部空污真的很嚴重，孩子常常咳嗽。' },
            { role: 'interviewer', text: '謝謝你花時間把想法說清楚。這些內容會被整理成逐字稿，納入審議分析。如果還有想補充的，隨時可以再開一輪。' },
          ],
        },
      ];
    }

    return [];
  }

  static saveConversations(sessionId: string, convs: HarmonicaConversation[]): void {
    try {
      localStorage.setItem(`${this.PREFIX}:conversations:${sessionId}`, JSON.stringify(convs));
    } catch (_) {}
  }

  static getParticipantConversation(sessionId: string, participantId: string): HarmonicaConversation | null {
    try {
      const raw = localStorage.getItem(`${this.PREFIX}:user_conv:${sessionId}:${participantId}`);
      if (raw) return JSON.parse(raw);
    } catch (_) {}
    return null;
  }

  static saveParticipantConversation(
    sessionId: string,
    participantId: string,
    conv: HarmonicaConversation
  ): void {
    try {
      localStorage.setItem(`${this.PREFIX}:user_conv:${sessionId}:${participantId}`, JSON.stringify(conv));
    } catch (_) {}

    // Also update overall conversations for this session
    const list = this.getConversations(sessionId);
    const idx = list.findIndex((c) => c.participant === conv.participant);
    if (idx >= 0) {
      list[idx] = conv;
    } else {
      list.push(conv);
    }
    this.saveConversations(sessionId, list);

    // Update session participant counters
    const session = this.getSession(sessionId);
    if (session) {
      session.participants = list.length;
      session.completed = list.filter((c) => c.done).length;
      this.saveSession(session);
    }
  }

  static getHostView(sessionId: string): HarmonicaHostView | null {
    const session = this.getSession(sessionId);
    if (!session) return null;
    const conversations = this.getConversations(sessionId);

    const summaries = conversations.map((c) => ({
      participant: c.participant,
      alias: c.alias,
      turns: c.turns,
      done: c.done,
      messages: c.messages.length,
      updatedAt: c.updatedAt,
      lastMessage: c.messages.length > 0 ? c.messages[c.messages.length - 1].text : '',
    }));

    return {
      ...session,
      conversations,
      summaries,
    };
  }
}
